pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -147,91 +203,445 @@ func (p *Pod) PodSets() []kueue.PodSet {
 			Name:  kueue.DefaultPodSetName,
 			Count: 1,
 			Template: corev1.PodTemplateSpec{
-				Spec: *p.Spec.DeepCopy(),
+				Spec: *p.pod.Spec.DeepCopy(),
 			},
 		},
 	}
 }
 
 // IsActive returns true if there are any running pods.
 func (p *Pod) IsActive() bool {
-	return p.Status.Phase == corev1.PodRunning
+	for i := range p.list.Items {
+		if p.list.Items[i].Status.Phase == corev1.PodRunning {
+			return true
+		}
+	}
+	return false
 }
 
-// PodsReady instructs whether job derived pods are all ready now.
-func (p *Pod) PodsReady() bool {
-	for i := range p.Status.Conditions {
-		c := &p.Status.Conditions[i]
+func hasPodReadyTrue(conds []corev1.PodCondition) bool {
+	for i := range conds {
+		c := conds[i]
 		if c.Type == corev1.PodReady {
 			return c.Status == corev1.ConditionTrue
 		}
 	}
 	return false
 }
 
+// PodsReady instructs whether job derived pods are all ready now.
+func (p *Pod) PodsReady() bool {
+	if !p.isGroup {
+		return hasPodReadyTrue(p.pod.Status.Conditions)
+	}
+
+	for i := range p.list.Items {
+		if !hasPodReadyTrue(p.list.Items[i].Status.Conditions) {
+			return false
+		}
+	}
+	return true
+}
+
 // GVK returns GVK (Group Version Kind) for the job.
 func (p *Pod) GVK() schema.GroupVersionKind {
 	return gvk
 }
 
-func (p *Pod) Stop(ctx context.Context, c client.Client, _ []podset.PodSetInfo, eventMsg string) (bool, error) {
-	// The podset info is not relevant here, since this should mark the pod's end of life
-	pCopy := &corev1.Pod{
-		ObjectMeta: metav1.ObjectMeta{
-			UID:       p.UID,
-			Name:      p.Name,
-			Namespace: p.Namespace,
-		},
-		TypeMeta: p.TypeMeta,
-		Status: corev1.PodStatus{
-			Conditions: []corev1.PodCondition{
-				{
-					Type:   ConditionTypeTerminationTarget,
-					Status: corev1.ConditionTrue,
-					LastTransitionTime: metav1.Time{
-						Time: time.Now(),
+func (p *Pod) Stop(ctx context.Context, c client.Client, wl *kueue.Workload, eventMsg string) (bool, error) {
+	var podsInGroup []corev1.Pod
+
+	if p.isGroup {
+		podsInGroup = p.list.Items
+	} else {
+		podsInGroup = []corev1.Pod{p.pod}
+	}
+
+	if wl != nil && p.isGroup {
+		if evCond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadEvicted); evCond != nil && evCond.Status == metav1.ConditionTrue {
```

Comment:
why do we need to check for WorkloadEvicted inside Stop? Wouldn't Stop only be called when the workload is evicted or unadmitted?

Reply:
Stop could be called when:
1. [wl has a deletion timestamp](https://github.com/kubernetes-sigs/kueue/blob/44ad0d50cacd1830c8ce2e50b1b584636d1e972a/pkg/controller/jobframework/reconciler.go#L234)
2. [job is unsuspended, but wl is not addmitted](https://github.com/kubernetes-sigs/kueue/blob/44ad0d50cacd1830c8ce2e50b1b584636d1e972a/pkg/controller/jobframework/reconciler.go#L355)
3. [job is unsuspended, but wl is not found](https://github.com/kubernetes-sigs/kueue/blob/44ad0d50cacd1830c8ce2e50b1b584636d1e972a/pkg/controller/jobframework/reconciler.go#L443)

Reply:
In all of those cases, we should issue Pod deletes, if the Pods don't already have DeletionTimestamps.

Reply:
That's what we're doing. On the other hand, if workload is evicted and all pods are stopped, or at least one pod is running, we don't stop the group.

Reply:
Ok, the logic is slightly off:

- If the wl is deleted or it doesn't exist: issue deletes for everything, if we didn't already.
- otherwise (eviction): delete anything that isn't suspended. There is no need to return an error if there is a running Pod, as we won't remove the reservation yet https://github.com/kubernetes-sigs/kueue/blob/44ad0d50cacd1830c8ce2e50b1b584636d1e972a/pkg/controller/jobframework/reconciler.go#L303